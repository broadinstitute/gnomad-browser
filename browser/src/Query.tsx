import React, { Component } from 'react'

import Delayed from './Delayed'
import StatusMessage from './StatusMessage'

const areVariablesEqual = (variables: any, otherVariables: any) => {
  const keys = Object.keys(variables)
  const otherKeys = Object.keys(otherVariables)
  if (keys.length !== otherKeys.length) {
    return false
  }
  return keys.every((key) => variables[key] === otherVariables[key])
}

const cancelable = (promise: any) => {
  let isCanceled = false
  const wrapper = new Promise((resolve: any, reject: any) => {
    promise.then(
      (value: any) => {
        if (!isCanceled) {
          resolve(value)
        }
      },
      (error: any) => {
        if (!isCanceled) {
          reject(error)
        }
      }
    )
  })

  return {
    cancel: () => {
      isCanceled = true
    },
    promise: wrapper,
  }
}

type BaseQueryState = any

type BaseQueryProps = {
  operationName: string | null
  query: string
  url: string
  variables?: any
  children: (state: BaseQueryState) => JSX.Element
}

export class BaseQuery extends Component<BaseQueryProps, BaseQueryState> {
  static defaultProps = {
    url: '/api/',
    operationName: null,
    variables: {},
  }

  currentRequest: any

  mounted: any

  state = {
    data: null,
    error: null,
    graphQLErrors: null,
    loading: true,
  }

  componentDidMount() {
    this.mounted = true
    this.loadData()
  }

  componentDidUpdate(prevProps: BaseQueryProps) {
    const { query, variables } = this.props
    if (query !== prevProps.query || !areVariablesEqual(variables, prevProps.variables)) {
      this.loadData()
    }
  }

  componentWillUnmount() {
    this.mounted = false
  }

  loadData() {
    const { operationName, query, url, variables } = this.props

    this.setState({
      loading: true,
      error: null,
      graphQLErrors: null,
    })

    if (this.currentRequest) {
      this.currentRequest.cancel()
    }

    this.currentRequest = cancelable(
      fetch(url, {
        body: JSON.stringify({
          operationName,
          query,
          variables,
        }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
      }).then((response) => response.json())
    )
    this.currentRequest.promise.then(
      (response: any) => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: response.data,
          error: null,
          graphQLErrors: response.errors,
          loading: false,
        })
      },
      (error: any) => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: null,
          error,
          graphQLErrors: null,
          loading: false,
        })
      }
    )
  }

  render() {
    const { children } = this.props
    return children(this.state)
  }
}

type OwnQueryProps = {
  children: (...args: any[]) => any
  errorMessage?: string
  loadingMessage?: string
  loadingPlaceholderHeight?: number
  operationName: string | null
  query: string
  success?: (...args: any[]) => any
  url: string
  variables?: any
}

// @ts-expect-error TS(2456) FIXME: Type alias 'QueryProps' circularly references itse... Remove this comment to see the full error message
type QueryProps = OwnQueryProps & typeof Query.defaultProps

// @ts-expect-error TS(7022) FIXME: 'Query' implicitly has type 'any' because it does ... Remove this comment to see the full error message
const Query = ({
  children,
  errorMessage,
  loadingMessage,
  loadingPlaceholderHeight,
  operationName,
  query,
  success,
  url,
  variables,
}: QueryProps) => {
  return (
    <BaseQuery operationName={operationName} query={query} url={url} variables={variables}>
      {({ data, error, graphQLErrors, loading }: any) => {
        if (loading) {
          return (
            <div style={{ height: loadingPlaceholderHeight || 'auto' }}>
              <Delayed>
                <StatusMessage>{loadingMessage}</StatusMessage>
              </Delayed>
            </div>
          )
        }

        if (error) {
          return <StatusMessage>{errorMessage}</StatusMessage>
        }

        if (!data || !success(data)) {
          return (
            <StatusMessage>
              {graphQLErrors && graphQLErrors.length
                ? Array.from(new Set(graphQLErrors.map((e: any) => e.message))).join(', ')
                : errorMessage}
            </StatusMessage>
          )
        }

        return children({ data })
      }}
    </BaseQuery>
  )
}

Query.defaultProps = {
  errorMessage: 'Error',
  loadingMessage: 'Loading',
  loadingPlaceholderHeight: undefined,
  success: () => true,
  url: '/api/',
  variables: {},
  operationName: null,
}

export default Query
