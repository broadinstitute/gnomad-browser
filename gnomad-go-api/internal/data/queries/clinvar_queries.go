package queries

import (
	"context"
	"encoding/json"
	"fmt"

	"gnomad-browser/gnomad-go-api/internal/elastic"
	"gnomad-browser/gnomad-go-api/internal/graph/model"
)

// ClinVar Elasticsearch indices
var ClinVarVariantIndices = map[string]string{
	"GRCh37": "clinvar_grch37_variants",
	"GRCh38": "clinvar_grch38_variants",
}

// ClinVarVariantDocument represents the ES document structure for ClinVar variants
type ClinVarVariantDocument struct {
	VariantID               string                   `json:"variant_id"`
	ReferenceGenome         string                   `json:"reference_genome"`
	Chrom                   string                   `json:"chrom"`
	Pos                     int                      `json:"pos"`
	Ref                     string                   `json:"ref"`
	Alt                     string                   `json:"alt"`
	ClinicalSignificance    string                   `json:"clinical_significance"`
	ClinVarVariationID      string                   `json:"clinvar_variation_id"`
	GoldStars               int                      `json:"gold_stars"`
	HGVSC                   *string                  `json:"hgvsc"`
	HGVSP                   *string                  `json:"hgvsp"`
	InGnomAD                *bool                    `json:"in_gnomad"`
	MajorConsequence        *string                  `json:"major_consequence"`
	ReviewStatus            string                   `json:"review_status"`
	TranscriptID            *string                  `json:"transcript_id"`
	RSID                    *string                  `json:"rsid"`
	LastEvaluated           *string                  `json:"last_evaluated"`
	TranscriptConsequences  []map[string]interface{} `json:"transcript_consequences"`
	GnomAD                  *ClinVarGnomADData       `json:"gnomad"`
	Submissions             []ClinVarSubmissionDoc   `json:"submissions"`
}

// ClinVarGnomADData represents gnomAD data within ClinVar variants
type ClinVarGnomADData struct {
	Exome  *ClinVarGnomADSequencingData `json:"exome"`
	Genome *ClinVarGnomADSequencingData `json:"genome"`
}

// ClinVarGnomADSequencingData represents sequencing type data within gnomAD
type ClinVarGnomADSequencingData struct {
	AC      int      `json:"ac"`
	AN      int      `json:"an"`
	Filters []string `json:"filters"`
}

// ClinVarSubmissionDoc represents a ClinVar submission
type ClinVarSubmissionDoc struct {
	ClinicalSignificance *string               `json:"clinical_significance"`
	LastEvaluated        *string               `json:"last_evaluated"`
	ReviewStatus         string                `json:"review_status"`
	SubmitterName        string                `json:"submitter_name"`
	Conditions           []ClinVarConditionDoc `json:"conditions"`
}

// ClinVarConditionDoc represents a ClinVar condition
type ClinVarConditionDoc struct {
	Name     string  `json:"name"`
	MedGenID *string `json:"medgen_id"`
}

// FetchClinVarVariant retrieves a ClinVar variant by variant ID and reference genome
func FetchClinVarVariant(ctx context.Context, client *elastic.Client, variantID string, referenceGenome string) (*model.ClinVarVariantDetails, error) {
	index, ok := ClinVarVariantIndices[referenceGenome]
	if !ok {
		return nil, fmt.Errorf("unsupported reference genome: %s", referenceGenome)
	}

	query := map[string]any{
		"query": map[string]any{
			"bool": map[string]any{
				"filter": map[string]any{
					"term": map[string]any{
						"variant_id": variantID,
					},
				},
			},
		},
		"_source": map[string]any{
			"includes": []string{"value"},
		},
		"size": 1,
	}

	response, err := client.Search(ctx, index, query)
	if err != nil {
		return nil, fmt.Errorf("elasticsearch search failed: %w", err)
	}

	if len(response.Hits.Hits) == 0 {
		return nil, nil
	}

	hit := response.Hits.Hits[0]

	// Extract the 'value' field from _source
	value, ok := hit.Source["value"].(map[string]any)
	if !ok {
		return nil, fmt.Errorf("unexpected document structure: missing 'value' field")
	}

	// Convert to JSON and back to parse into our struct
	valueJSON, err := json.Marshal(value)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal document: %w", err)
	}

	var doc ClinVarVariantDocument
	if err := json.Unmarshal(valueJSON, &doc); err != nil {
		return nil, fmt.Errorf("failed to unmarshal document: %w", err)
	}

	// Convert to GraphQL model
	result := &model.ClinVarVariantDetails{
		VariantID:            doc.VariantID,
		ReferenceGenome:      mapReferenceGenome(doc.ReferenceGenome),
		Chrom:                doc.Chrom,
		Pos:                  doc.Pos,
		Ref:                  doc.Ref,
		Alt:                  doc.Alt,
		ClinicalSignificance: doc.ClinicalSignificance,
		ClinvarVariationID:   doc.ClinVarVariationID,
		GoldStars:            doc.GoldStars,
		InGnomad:             doc.InGnomAD != nil && *doc.InGnomAD,
		ReviewStatus:         doc.ReviewStatus,
	}

	// Optional fields
	if doc.RSID != nil {
		result.Rsid = doc.RSID
	}
	if doc.LastEvaluated != nil {
		result.LastEvaluated = doc.LastEvaluated
	}

	// Convert gnomAD data - always return structure with null fields when no data
	if doc.GnomAD != nil {
		result.Gnomad = convertClinVarGnomADData(doc.GnomAD)
	} else {
		result.Gnomad = &model.ClinVarVariantGnomadData{
			Exome:  nil,
			Genome: nil,
		}
	}

	// Convert submissions
	if len(doc.Submissions) > 0 {
		result.Submissions = make([]*model.ClinVarSubmission, len(doc.Submissions))
		for i, sub := range doc.Submissions {
			result.Submissions[i] = convertClinVarSubmission(sub)
		}
	}

	return result, nil
}

// convertClinVarGnomADData converts ClinVar gnomAD data to GraphQL model
func convertClinVarGnomADData(data *ClinVarGnomADData) *model.ClinVarVariantGnomadData {
	result := &model.ClinVarVariantGnomadData{}

	if data.Exome != nil && (data.Exome.AC > 0 || data.Exome.AN > 0 || len(data.Exome.Filters) > 0) {
		result.Exome = &model.ClinVarVariantGnomadSequencingTypeData{
			Ac:      data.Exome.AC,
			An:      data.Exome.AN,
			Filters: data.Exome.Filters,
		}
	}

	if data.Genome != nil && (data.Genome.AC > 0 || data.Genome.AN > 0 || len(data.Genome.Filters) > 0) {
		result.Genome = &model.ClinVarVariantGnomadSequencingTypeData{
			Ac:      data.Genome.AC,
			An:      data.Genome.AN,
			Filters: data.Genome.Filters,
		}
	}

	return result
}

// convertClinVarSubmission converts a ClinVar submission to GraphQL model
func convertClinVarSubmission(sub ClinVarSubmissionDoc) *model.ClinVarSubmission {
	result := &model.ClinVarSubmission{
		ReviewStatus:  sub.ReviewStatus,
		SubmitterName: sub.SubmitterName,
	}

	// Optional fields
	if sub.ClinicalSignificance != nil {
		result.ClinicalSignificance = sub.ClinicalSignificance
	}
	if sub.LastEvaluated != nil {
		result.LastEvaluated = sub.LastEvaluated
	}

	// Convert conditions
	if len(sub.Conditions) > 0 {
		result.Conditions = make([]*model.ClinVarCondition, len(sub.Conditions))
		for i, cond := range sub.Conditions {
			result.Conditions[i] = &model.ClinVarCondition{
				Name: cond.Name,
			}
			if cond.MedGenID != nil {
				result.Conditions[i].MedgenID = cond.MedGenID
			}
		}
	}

	return result
}

// mapReferenceGenome maps string to enum
func mapReferenceGenome(genome string) model.ReferenceGenomeID {
	switch genome {
	case "GRCh37":
		return model.ReferenceGenomeIDGRCh37
	case "GRCh38":
		return model.ReferenceGenomeIDGRCh38
	default:
		return model.ReferenceGenomeIDGRCh38 // default
	}
}

// FetchClinVarVariantsByGene fetches ClinVar variants for a gene
func FetchClinVarVariantsByGene(ctx context.Context, client *elastic.Client, geneID string, referenceGenome string) ([]*model.ClinVarVariant, error) {
	index, ok := ClinVarVariantIndices[referenceGenome]
	if !ok {
		return nil, fmt.Errorf("unsupported reference genome: %s", referenceGenome)
	}

	query := map[string]any{
		"query": map[string]any{
			"bool": map[string]any{
				"filter": map[string]any{
					"term": map[string]any{
						"gene_id": geneID,
					},
				},
			},
		},
		"_source": map[string]any{
			"includes": []string{"value"},
		},
		"sort": []map[string]any{
			{"pos": map[string]any{"order": "asc"}},
		},
		"size": 10000,
	}

	response, err := client.Search(ctx, index, query)
	if err != nil {
		return []*model.ClinVarVariant{}, fmt.Errorf("elasticsearch search failed: %w", err)
	}

	// Create gene context for transcript consequence selection
	context := FlagContext{
		Type:   "gene",
		GeneID: geneID,
	}

	results := make([]*model.ClinVarVariant, 0, len(response.Hits.Hits))
	for _, hit := range response.Hits.Hits {
		variant, err := convertHitToClinVarVariantWithContext(hit, &context)
		if err != nil {
			continue // Skip invalid variants
		}
		results = append(results, variant)
	}

	return results, nil
}

// FetchClinVarVariantsByRegion fetches ClinVar variants for a region
func FetchClinVarVariantsByRegion(ctx context.Context, client *elastic.Client, chrom string, start, stop int, referenceGenome string) ([]*model.ClinVarVariant, error) {
	index, ok := ClinVarVariantIndices[referenceGenome]
	if !ok {
		return nil, fmt.Errorf("unsupported reference genome: %s", referenceGenome)
	}

	query := map[string]any{
		"query": map[string]any{
			"bool": map[string]any{
				"filter": []map[string]any{
					{"term": map[string]any{"chrom": chrom}},
					{"range": map[string]any{
						"pos": map[string]any{
							"gte": start,
							"lte": stop,
						},
					}},
				},
			},
		},
		"_source": map[string]any{
			"includes": []string{"value"},
		},
		"sort": []map[string]any{
			{"pos": map[string]any{"order": "asc"}},
		},
		"size": 10000,
	}

	response, err := client.Search(ctx, index, query)
	if err != nil {
		return []*model.ClinVarVariant{}, fmt.Errorf("elasticsearch search failed: %w", err)
	}

	// Create region context for transcript consequence selection
	context := FlagContext{
		Type: "region",
	}

	results := make([]*model.ClinVarVariant, 0, len(response.Hits.Hits))
	for _, hit := range response.Hits.Hits {
		variant, err := convertHitToClinVarVariantWithContext(hit, &context)
		if err != nil {
			continue // Skip invalid variants
		}
		results = append(results, variant)
	}

	return results, nil
}

// FetchClinVarVariantsByTranscript fetches ClinVar variants for a transcript
func FetchClinVarVariantsByTranscript(ctx context.Context, client *elastic.Client, transcriptID string, referenceGenome string) ([]*model.ClinVarVariant, error) {
	index, ok := ClinVarVariantIndices[referenceGenome]
	if !ok {
		return nil, fmt.Errorf("unsupported reference genome: %s", referenceGenome)
	}

	query := map[string]any{
		"query": map[string]any{
			"bool": map[string]any{
				"filter": map[string]any{
					"term": map[string]any{
						"transcript_id": transcriptID,
					},
				},
			},
		},
		"_source": map[string]any{
			"includes": []string{"value"},
		},
		"sort": []map[string]any{
			{"pos": map[string]any{"order": "asc"}},
		},
		"size": 10000,
	}

	response, err := client.Search(ctx, index, query)
	if err != nil {
		return []*model.ClinVarVariant{}, fmt.Errorf("elasticsearch search failed: %w", err)
	}

	results := make([]*model.ClinVarVariant, 0, len(response.Hits.Hits))
	
	// Create transcript context for proper field extraction
	context := FlagContext{
		Type:         "transcript",
		TranscriptID: transcriptID,
	}
	
	for _, hit := range response.Hits.Hits {
		variant, err := convertHitToClinVarVariantWithContext(hit, &context)
		if err != nil {
			continue // Skip invalid variants
		}
		results = append(results, variant)
	}

	return results, nil
}

// convertHitToClinVarVariant converts an Elasticsearch hit to a ClinVarVariant model
func convertHitToClinVarVariant(hit elastic.Hit) (*model.ClinVarVariant, error) {
	// Use empty context for backward compatibility
	return convertHitToClinVarVariantWithContext(hit, nil)
}

// convertHitToClinVarVariantWithContext converts an Elasticsearch hit to a ClinVarVariant model with context
func convertHitToClinVarVariantWithContext(hit elastic.Hit, context *FlagContext) (*model.ClinVarVariant, error) {
	// Extract the 'value' field from _source
	value, ok := hit.Source["value"].(map[string]any)
	if !ok {
		return nil, fmt.Errorf("unexpected document structure: missing 'value' field")
	}

	// Convert to JSON and back to parse into our struct
	valueJSON, err := json.Marshal(value)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal document: %w", err)
	}

	var doc ClinVarVariantDocument
	if err := json.Unmarshal(valueJSON, &doc); err != nil {
		return nil, fmt.Errorf("failed to unmarshal document: %w", err)
	}

	// Convert to GraphQL model
	result := &model.ClinVarVariant{
		VariantID:            doc.VariantID,
		ReferenceGenome:      mapReferenceGenome(doc.ReferenceGenome),
		Chrom:                doc.Chrom,
		Pos:                  doc.Pos,
		Ref:                  doc.Ref,
		Alt:                  doc.Alt,
		ClinicalSignificance: doc.ClinicalSignificance,
		ClinvarVariationID:   doc.ClinVarVariationID,
		GoldStars:            doc.GoldStars,
		ReviewStatus:         doc.ReviewStatus,
	}

	// Optional fields
	if doc.HGVSC != nil {
		result.Hgvsc = doc.HGVSC
	}
	if doc.HGVSP != nil {
		result.Hgvsp = doc.HGVSP
	}
	if doc.InGnomAD != nil {
		result.InGnomad = doc.InGnomAD
	}

	// Handle transcript consequences with context
	if context != nil && len(doc.TranscriptConsequences) > 0 {
		// Use GetConsequenceForContext to get the appropriate transcript consequence
		if consequence := GetConsequenceForContext(doc.TranscriptConsequences, *context); consequence != nil {
			// Extract major_consequence and transcript_id from the selected consequence
			if majorConsequence, ok := consequence["major_consequence"].(string); ok {
				result.MajorConsequence = &majorConsequence
			}
			if transcriptID, ok := consequence["transcript_id"].(string); ok {
				result.TranscriptID = &transcriptID
			}
		}
	} else {
		// Fall back to document-level fields for backward compatibility
		if doc.MajorConsequence != nil {
			result.MajorConsequence = doc.MajorConsequence
		}
		if doc.TranscriptID != nil {
			result.TranscriptID = doc.TranscriptID
		}
	}

	// Convert gnomAD data - always return structure with null fields when no data
	if doc.GnomAD != nil {
		result.Gnomad = convertClinVarGnomADData(doc.GnomAD)
	} else {
		result.Gnomad = &model.ClinVarVariantGnomadData{
			Exome:  nil,
			Genome: nil,
		}
	}

	return result, nil
}

