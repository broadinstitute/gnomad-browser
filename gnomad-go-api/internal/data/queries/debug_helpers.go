package queries

import (
	"encoding/json"
	"fmt"
)

// DebugPopulationData prints the raw population data for debugging
func DebugPopulationData(popData []map[string]interface{}, label string) {
	fmt.Printf("\n=== DEBUG: %s ===\n", label)
	for _, pop := range popData {
		id := pop["id"]
		ac := pop["ac"]
		an := pop["an"]
		homCount := pop["homozygote_count"]
		hemiCount := pop["hemizygote_count"]
		
		// Only print populations with non-zero homozygote counts
		if homCount != nil && homCount.(float64) > 0 {
			fmt.Printf("Population %s: ac=%v, an=%v, homozygote_count=%v, hemizygote_count=%v\n",
				id, ac, an, homCount, hemiCount)
		}
	}
	
	// Also print full JSON for inspection
	data, _ := json.MarshalIndent(popData, "", "  ")
	fmt.Printf("\nFull population data:\n%s\n", string(data))
}