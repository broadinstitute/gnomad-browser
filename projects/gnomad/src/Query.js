import gqlFetch from 'graphql-fetch'
import PropTypes from 'prop-types'
import { Component } from 'react'

const areVariablesEqual = (variables, otherVariables) => {
  const keys = Object.keys(variables)
  const otherKeys = Object.keys(otherVariables)
  if (keys.length !== otherKeys.length) {
    return false
  }
  return keys.every(key => variables[key] === otherVariables[key])
}

const cancelable = promise => {
  let isCanceled = false
  const wrapper = new Promise((resolve, reject) => {
    promise.then(
      value => {
        if (!isCanceled) {
          resolve(value)
        }
      },
      error => {
        if (!isCanceled) {
          reject(error)
        }
      }
    )
  })

  return {
    cancel: () => {
      isCanceled = true
    },
    promise: wrapper,
  }
}

export class Query extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
    variables: PropTypes.object,
  }

  static defaultProps = {
    variables: {},
  }

  state = {
    data: null,
    error: null,
    loading: true,
  }

  componentDidMount() {
    this.mounted = true
    this.loadData()
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.query !== prevProps.query ||
      !areVariablesEqual(this.props.variables, prevProps.variables)
    ) {
      this.loadData()
    }
  }

  componentWillUnmount() {
    this.mounted = false
  }

  loadData() {
    this.setState({
      loading: true,
      error: null,
    })

    if (this.currentRequest) {
      this.currentRequest.cancel()
    }

    this.currentRequest = cancelable(
      gqlFetch(process.env.GNOMAD_API_URL)(this.props.query, this.props.variables)
    )
    this.currentRequest.promise.then(
      response => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: response.data,
          error: null,
          loading: false,
        })
      },
      error => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: null,
          error,
          loading: false,
        })
      }
    )
  }

  render() {
    return this.props.children(this.state)
  }
}
