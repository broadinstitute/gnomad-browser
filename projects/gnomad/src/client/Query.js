import gqlFetch from 'graphql-fetch'
import PropTypes from 'prop-types'
import { Component } from 'react'

const areVariablesEqual = (variables, otherVariables) => {
  const keys = Object.keys(variables)
  const otherKeys = Object.keys(otherVariables)
  if (keys.length !== otherKeys.length) {
    return false
  }
  return keys.every(key => variables[key] === otherVariables[key])
}

const cancelable = promise => {
  let isCanceled = false
  const wrapper = new Promise((resolve, reject) => {
    promise.then(
      value => {
        if (!isCanceled) {
          resolve(value)
        }
      },
      error => {
        if (!isCanceled) {
          reject(error)
        }
      }
    )
  })

  return {
    cancel: () => {
      isCanceled = true
    },
    promise: wrapper,
  }
}

class Query extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
    url: PropTypes.string,
    variables: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  }

  static defaultProps = {
    url: '/api/',
    variables: {},
  }

  state = {
    data: null,
    error: null,
    graphQLErrors: null,
    loading: true,
  }

  componentDidMount() {
    this.mounted = true
    this.loadData()
  }

  componentDidUpdate(prevProps) {
    const { query, variables } = this.props
    if (query !== prevProps.query || !areVariablesEqual(variables, prevProps.variables)) {
      this.loadData()
    }
  }

  componentWillUnmount() {
    this.mounted = false
  }

  loadData() {
    const { query, url, variables } = this.props

    this.setState({
      loading: true,
      error: null,
      graphQLErrors: null,
    })

    if (this.currentRequest) {
      this.currentRequest.cancel()
    }

    this.currentRequest = cancelable(gqlFetch(url)(query, variables))
    this.currentRequest.promise.then(
      response => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: response.data,
          error: null,
          graphQLErrors: response.errors,
          loading: false,
        })
      },
      error => {
        if (!this.mounted) {
          return
        }
        this.setState({
          data: null,
          error,
          graphQLErrors: null,
          loading: false,
        })
      }
    )
  }

  render() {
    const { children } = this.props
    return children(this.state)
  }
}

export default Query
