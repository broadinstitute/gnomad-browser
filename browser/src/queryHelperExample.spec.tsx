// There is a certain amount of boilerplate involved in using the query
// test helpers, so this file exists as an annotated working example.
//
// Start like you would any test:
import React from 'react'
import renderer from 'react-test-renderer'

// These next imports must appear in this order
import { jest, expect } from '@jest/globals'
import { mockQueries } from '../../tests/__helpers__/queries'
import Query, { BaseQuery } from './Query'

// You'd think that you can wrap this in a helper function in queries.tsx,
// but it won't work if you do.
jest.mock('./Query', () => {
  const originalModule = jest.requireActual('./Query')

  return {
    __esModule: true,
    ...(originalModule as object),
    default: jest.fn(),
    BaseQuery: jest.fn(),
  }
})

// This is the API that your tests will use to interact with the mock queries
const { resetMockApiCalls, resetMockApiResponses, simulateApiResponse, setMockApiResponses } =
  mockQueries()

// You can do this on a test-by-test basis too, rather than beforeEach.
beforeEach(() => {
  Query.mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('Query', query, children, operationName, variables)
    )
  )
  // The semicolon is due to a quirk of JS/TS parsing--try taking it out and
  // you'll see what happens.
  // Also, it's not clear why we have to cast BaseQuery here but not Query
  // above.
  ;(BaseQuery as any).mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('BaseQuery', query, children, operationName, variables)
    )
  )
})

// Mock query calls are not cleaned up automatically
afterEach(() => {
  resetMockApiCalls()
  resetMockApiResponses()
})

// Now for the example code proper.

const ExampleInnerComponent = ({ content }: { content: any }) => (
  <>
    My content is as follows:
    <pre>{JSON.stringify(content)}</pre>
  </>
)

// The following is a simplified version, with no loading state or error
// handling, of the Query/BaseQuery boilerplate that's copy/pasted throughout
// the code.
const ExampleQuery = () => (
  <BaseQuery query="fakeQuery" operationName="exampleOperation" variables={{}}>
    {({ data }: any) => {
      return <ExampleInnerComponent content={data.content!} />
    }}
  </BaseQuery>
)

const ExampleOuterComponent = () => (
  <div>
    The fake query results go here: <ExampleQuery />
  </div>
)

describe('query helpers', () => {
  test('get out the data you put in', () => {
    // The argument to setMockApiResponses is keyed by operation name
    setMockApiResponses({
      exampleOperation: () => ({
        content: ['a', { one: 'two' }, 999],
      }),
    })

    const tree = renderer.create(<ExampleOuterComponent />)
    expect(tree).toMatchSnapshot()
  })
})
