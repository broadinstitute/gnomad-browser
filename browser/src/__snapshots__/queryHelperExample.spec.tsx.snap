// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query helpers get out the data you put in 1`] = `
<div>
  The fake query results go here: 
  My content is as follows:
  <pre>
    ["a",{"one":"two"},999]
  </pre>
</div>
`;
